### ISSUE TITLE
Bug with System.Linq.Enumerable.Average<Int64>(this IEnumerable<long> source) method

### ISSUE THREAD

**Issue by fengzhenqiong** [dotnet/corefx#40620]
Seems there's something wrong with the `System.Linq.Enumerable.Average<long>(this IEnumerable<long> source)` method.

Let's say, if I have a collection with only two large numbers (larger than 2^62), if I call this method it will throw System.OverflowException saying that "Arithmetic operation resulted in an overflow".

I used the following code:
`
var numbers = new[] {8000000000000000002L, 7999999999999999998L};
Console.WriteLine(numbers.Average());
`

I also had a look at the source code in .NET Framework 4.8 and seems this problem is there, too. Is this a bug (I think it is), or do you have other considerations?

Anyway, fixing this problem is quite straightforward so I think maybe we should get rid of it.

Thanks.

**Comment by fengzhenqiong** [comment]
Seems there are also other methods with this same problem.

**Comment by fengzhenqiong** [comment]
Yup, this is a little confusing as Int32 collections don't have such problem, and the return value is double and promising for Int64.

The most important is that this Overflow is not necessary.

**Comment by fengzhenqiong** [comment]
> > The most important is that this Overflow is not necessary.
> 
> What implementation would you propose to avoid it that is also accurate and doesn't negatively impact performance significantly?

Implementation is straightforward by simply replacing the sum with Decimal, but speed is around 5 times slower with .NET Core 2.2 and 2~3 times slower with .NET Framework 4.8. I just did this simple test on my laptop.

Double gives as good performance almost the same as Int64, but the result is not accurate with large numbers(as it uses scientific numbers).

I also tried to implement it myself with Int64 but performance is more or less the same with Decimal scenario on .NET Core 2.2 as it perform several multiply/divide operation for every number.

So it seems this is something relates to the decision making. I prefer functionality to performance (as we can improve this during framework evolution). I come into this issue when I try to average several DateTime.Ticks values in my test codes. If there is OverflowException, it make this method of less use.

Anyway, as it's already in the documentation, so it may not be a bug, but a known issue.

Or, maybe you guys can provide another overload of this method to provide the full functionality, but with slower speed? Just some random thought.

All in all, I really love .NET Core a lot and wish it becomes better and better. Thanks a lot.

**Comment by fengzhenqiong** [comment]
Yup, averaging the Ticks makes little sense and that was just some test codes. The Overflow interests me because usually the sum of a collection of numbers will likely exceeds the MaxValue limit.

Int32 avoid such problem with Int64. Int128 will solve this problem for Int64, but it will also bring the same problem averaging Int128.  I think the Convert to Decimal approach is good enough here.

Thanks a lot for the detailed explanation.

**Comment by fengzhenqiong** [comment]
> > decimal should be fine up to 29 digits, but will come at a perf cost.
> 
> Just switching to `decimal sum` incurs a 4x perf hit on my machine. [stephentoub@5c33656](https://github.com/stephentoub/corefx/commit/5c33656afafc445d467b43c91c21ecbef229a0e7), which switches to decimal on overflow, incurs an ~15% perf hit on my machine.

This good compromise got my vote for its good enough performance and exact accuracy. :smile:

**Comment by GSPP** [comment]
Callers can `myLongs.Select(x => (double)x).Average()` as a workaround. No new APIs are needed.

**Comment by tannergooding** [comment]
> Yup, this is a little confusing as Int32 collections don't have such problem

It should be clarified that `Int32` collections do have this problem. You just need to have an enumeration that returns `int.MaxValue` ~4 billion times (which also doesn't take terribly long to execute on modern computers).

At the end of the day, I think it would be better to expose an explicit API which is documented to handle the overflow scenario using `BigInteger`.

**Comment by stephentoub** [comment]
> It should be clarified that Int32 collections do have this problem. You just need to have an enumeration that returns int.MaxValue ~4 billion times (which also doesn't take terribly long to execute on modern computers).

In practice, there's a big difference between having an `IEnumerable<int>` with 4294967296 elements and an `IEnumerable<long>` with 2.

**Comment by tannergooding** [comment]
Sure, but at the same time you'll be able to overflow `decimal` after only ~1b elements; and if `Int128` ends up being supported for .NET 5 you'll have the same problem with `Int128` (can potentially overflow after 2 elements).

It would likely be better to just have a single API that handles this case (and any existing case) for when that is interesting to the user.

There are also other problems when dealing with numbers that large as well, such as not have a result which can be accurately returned. `double` can only represent whole integers up to 2^53. By the time you are providing inputs at or above 2^53, the gap between representable values is `1` and doubles every power of two (gap is `2` at 2^54, `4` at 2^55, and the gap is `1024` at 2^64). This only increases the problem as users provide larger inputs and the API no longer functions as intended.

**Comment by stephentoub** [comment]
> It would likely be better to just have a single API that handles this case (and any existing case) for when that is interesting to the user.

I disagree.  We have an API that claims to support averaging an `IEnumerable<long>`, a data type that's been in .NET for 20 years, but it can't handle an enumerable with only two inputs.  How many arrays have you seen with two inputs vs with ~1b elements?  The scenarios we're talking about here are very different; we should not use a super corner case to justify the behavior for one that's much more mainstream.

> double can only represent whole integers up to 2^53

Right... that's why the workaround I proposed doesn't use double.

**Comment by tannergooding** [comment]
> Right... that's why the workaround I proposed doesn't use double.

This isn't an issue with using `double` to compute the intermediate sum; this is an issue with returning `double` as the result that will only be exacerbated by allowing users to pass in even larger values (and we can't change the result from returning `double`).

**Comment by stephentoub** [comment]
I'm honestly not understanding what you're arguing for.  Are you saying this Average method should have never existed, should never be used, should be deprecated, etc.?

With this API, there will always exist the potential for a skewed result due to it returning a `double`; that's the nature of any API which casts a `long` to a `double`.  That doesn't mean it's acceptable for the method to throw overflow exceptions in common cases.

**Comment by tannergooding** [comment]
I'm just trying to ensure the tradeoffs of fixing the bug, for the given API, are well understood. The maximum error today is limited by the fact that you can't have a base sum over `long.MaxValue`, extending this will increase the maximum error possible for the result returned from the API and will likely just better expose an existing issue with the API that people will have to deal with instead.

For example, lets assume you are summing `long.MaxValue - 10` through `long.MaxValue`; this equals `101457092405402533822` and the average is `9223372036854775802`.

However, the nearest representable `double` to the sum is `101457092405402533888` (off by 66) and `(double)sum / count` (to compute the nearest representable average) then ends up being `9223372036854775808` (off by 6).

This rounding error will only increase as larger sums become allowed for the given algorithm and will quickly become a problem for anyone dealing with large `long` (e.g. summing `long.MaxValue - 1000` through `long.MaxValue` ends up being off by over 500).

**Comment by tannergooding** [comment]
There are other potential mitigations to help reduce the rounding error (you could compute the average as `decimal` and then `return (double)decimalAvg;`); but they come with trade-offs such as potentially changing the existing result returned for certain sums (and in some cases, they won't change the result at all because the nearest value is up to 1024 away).

**Comment by fengzhenqiong** [comment]
If use Decimal fallback approach the `Int32` Average can get rid of this issue as `Int32.MaxValue * Int64.MaxValue` is less than `Decimal.MaxValue`.

But the `Int64` version Average still have this problem.
